### DISCUSSION TITLE
Feature request: Github App permissions to issues/PRs is a security hole

### DISCUSSION THREAD

**Discussion by sandstrom** [community/community#63901] [Apps, API and Webhooks]
### Select Topic Area

Product Feedback

### Body

When a third-party Github application gets hacked (the recent [Heroku hack](https://www.theregister.com/2022/04/21/github-stolen-oauth-tokens-used-in-breaches/) for example, that leaked Github credentials) the permissions granted matter.

A good trend with integrations is that many providers (Readme.io and Sentry.io are two examples) will use a CLI tool in the Github Actions runner, and this tool will post data to Readme/Sentry backends. That way, their Github Apps needs less permissions (no full-repo access).

And the CLI tool itself can be SHA-pinned, via NPM or similar.

However, many of them still require PR/Issue access, to update PRs with information and to set a status.

It would be good if Github could better accommodate this use-case, therefore allowing apps to run with fewer permissions, so a hacked Github app causes fewer problems.

This could be:

- Time-limits on permission for PRs/issues, only those open or closed within the past 12 months
- Better support for Github Actions runners to set status, update PR text and otherwise supplant a PR with metadata
  - For example, it could be a pre-signed and time-limited URL that the GH-runner could create (assuming that it has permissions itself), for updating the PR. The runner would then pass this on to the backend, which would use it for an async POST of a comment/status later.

**Comment by Mronyx** [comment]
Make sure to replace {owner} and {repo} with the respective values for your repository. Additionally, you may need to include any required authentication headers depending on the API's authentication requirements.

This code snippet sends a GET request to the GitHub API's endpoint for listing branches, with the head parameter set to "master" (or any other branch name you specify) and the timestamp parameter set to the current timestamp. The response is then processed accordingly, based on the response status code.

**Reply by sandstrom** [reply]
Are you answering in the right thread?

This isn't a question about how the API works, it's feedback.

**Comment by Mronyx** [comment]
To solve this problem, GitHub could enhance its permissions system to allow more granular control over the access granted to third-party applications. This way, developers would have the flexibility to grant specific permissions based on their needs, reducing the risk and potential impact of a hacked GitHub app.

One possible solution could involve introducing new permission scopes that specifically cater to the use-case mentioned in the description. For example, GitHub could introduce a separate permission scope for updating pull requests and setting statuses, without granting the app full-repo access.

Additionally, GitHub could improve its tools and documentation to guide developers on best practices for securing their integrations. They could provide recommendations on minimizing the permissions required for an app's functionality, suggesting the use of CLI tools or other mechanisms to access necessary data without requiring full access to the repository.

GitHub could also consider implementing stricter security measures for third-party apps, such as mandatory multifactor authentication for app owners and regular security audits to identify potential vulnerabilities.

Overall, by offering more control and guidance to developers, GitHub can help mitigate the risks associated with hacked GitHub apps and minimize the potential damage caused by such incidents.
